x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/FoldContext.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.core.expression;
+
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.breaker.CircuitBreakingException;
+import org.elasticsearch.xpack.esql.core.QlClientException;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+
+/**
+ * Context passed to {@link Expression#fold}.
+ */
+public class FoldContext {
+    /**
+     * {@link Expression#fold} using any amount of memory. Only safe for tests.
+     */
+    public static FoldContext unbounded() {
+        return new FoldContext(Long.MAX_VALUE);
+    }
+
+    /**
+     * {@link Expression#fold} using a small amount of memory.
+     */
+    public static FoldContext small() {
```

Comment:
Are we using this? It's unbounded

Reply:
I'll double check before merging. We really should only have one unbounded one.

Reply:
At some point I was playing around with the idea that you could know up front that a fold would always be "small" so you didn't have to limit it. But that's:
1. Impossible to know
2. Weird to want to do. If you know it's small them limit it.